File: backend/models/transport.py
"""
Transport Models
Database models for transport/shipment management
"""
from sqlalchemy import Column, String, Numeric, ForeignKey, Text, Date, Enum as SQLEnum
from sqlalchemy.orm import relationship
from sqlalchemy.dialects.postgresql import UUID
from datetime import date
from enum import Enum as PyEnum
import uuid

from models.base import BaseModel


class VehicleType(str, PyEnum):
    """Vehicle type enumeration"""

    CAR = "car"
    VAN = "van"
    TRUCK = "truck"
    MOTORCYCLE = "motorcycle"
    OTHER = "other"


class VehicleStatus(str, PyEnum):
    """Vehicle status enumeration"""

    ACTIVE = "active"
    MAINTENANCE = "maintenance"
    INACTIVE = "inactive"


class ShipmentStatus(str, PyEnum):
    """Shipment status enumeration"""

    PENDING = "pending"  # Created, not started
    IN_TRANSIT = "in_transit"  # En ruta
    DELIVERED = "delivered"  # Entregado
    CANCELLED = "cancelled"  # Cancelado


class ExpenseType(str, PyEnum):
    """Shipment expense type"""

    FUEL = "fuel"  # Combustible
    TOLL = "toll"  # Peajes
    PARKING = "parking"  # Estacionamiento
    MAINTENANCE = "maintenance"  # Mantenimiento
    OTHER = "other"  # Otros


class Vehicle(BaseModel):
    """
    Vehicle Model
    Represents a vehicle in the fleet
    """

    __tablename__ = "vehicles"

    # Multi-tenancy
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False)

    # Basic Information
    plate_number = Column(
        String(20), nullable=False, index=True, comment="License plate number"
    )
    brand = Column(String(100), nullable=False, comment="Vehicle brand (e.g., Toyota)")
    model = Column(String(100), nullable=False, comment="Vehicle model")
    year = Column(String(4), nullable=True, comment="Manufacturing year")
    vehicle_type = Column(
        SQLEnum(VehicleType),
        nullable=False,
        default=VehicleType.CAR,
        comment="Type of vehicle",
    )

    # Status and Assignment
    status = Column(
        SQLEnum(VehicleStatus),
        nullable=False,
        default=VehicleStatus.ACTIVE,
        index=True,
        comment="Current status",
    )
    assigned_driver_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True,
        comment="Assigned driver user_id",
    )

    # Specifications
    capacity_kg = Column(
        Numeric(10, 2), nullable=True, comment="Load capacity in kilograms"
    )
    fuel_type = Column(String(50), nullable=True, comment="Fuel type (Gasolina, Diesel, Eléctrico)")
    fuel_efficiency_km_l = Column(
        Numeric(5, 2), nullable=True, comment="Fuel efficiency in km per liter"
    )

    # Maintenance
    last_maintenance_date = Column(Date, nullable=True, comment="Last maintenance date")
    next_maintenance_date = Column(
        Date, nullable=True, comment="Next scheduled maintenance"
    )
    mileage_km = Column(Numeric(10, 2), nullable=True, comment="Current mileage in km")

    # Notes
    notes = Column(Text, nullable=True, comment="Additional notes")

    # Relationships
    driver = relationship("User", foreign_keys=[assigned_driver_id])
    shipments = relationship("Shipment", back_populates="vehicle")

    def __repr__(self):
        return f"<Vehicle {self.plate_number} - {self.brand} {self.model}>"


class Shipment(BaseModel):
    """
    Shipment Model
    Represents a freight/shipment delivery
    """

    __tablename__ = "shipments"

    # Multi-tenancy
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False)

    # Shipment Information
    shipment_number = Column(
        String(50),
        nullable=False,
        unique=True,
        index=True,
        comment="Unique shipment number",
    )
    client_id = Column(
        UUID(as_uuid=True),
        ForeignKey("clients.id"),
        nullable=True,
        comment="Client receiving the shipment",
    )
    vehicle_id = Column(
        UUID(as_uuid=True),
        ForeignKey("vehicles.id"),
        nullable=True,
        comment="Assigned vehicle",
    )
    driver_id = Column(
        UUID(as_uuid=True),
        ForeignKey("users.id"),
        nullable=True,
        comment="Assigned driver",
    )

    # Locations
    origin_address = Column(String(255), nullable=False, comment="Origin address")
    origin_city = Column(String(100), nullable=False, comment="Origin city")
    destination_address = Column(
        String(255), nullable=False, comment="Destination address"
    )
    destination_city = Column(String(100), nullable=False, comment="Destination city")

    # Dates
    scheduled_date = Column(
        Date, nullable=False, comment="Scheduled pickup/delivery date"
    )
    pickup_date = Column(Date, nullable=True, comment="Actual pickup date")
    delivery_date = Column(Date, nullable=True, comment="Actual delivery date")

    # Cargo Details
    description = Column(Text, nullable=True, comment="Cargo description")
    weight_kg = Column(Numeric(10, 2), nullable=True, comment="Weight in kg")
    quantity = Column(Numeric(10, 2), nullable=True, comment="Quantity of items")

    # Costs
    estimated_distance_km = Column(
        Numeric(10, 2), nullable=True, comment="Estimated distance in km"
    )
    actual_distance_km = Column(
        Numeric(10, 2), nullable=True, comment="Actual distance traveled"
    )
    freight_cost = Column(
        Numeric(12, 2), nullable=False, default=0, comment="Freight cost charged"
    )
    currency = Column(String(3), nullable=False, default="USD", comment="Currency")

    # Status
    status = Column(
        SQLEnum(ShipmentStatus),
        nullable=False,
        default=ShipmentStatus.PENDING,
        index=True,
        comment="Current status",
    )

    # Notes
    notes = Column(Text, nullable=True, comment="Additional notes")

    # Relationships
    client = relationship("Client", foreign_keys=[client_id])
    vehicle = relationship("Vehicle", back_populates="shipments")
    driver = relationship("User", foreign_keys=[driver_id])
    expenses = relationship("ShipmentExpense", back_populates="shipment", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<Shipment {self.shipment_number} - {self.status.value}>"


class ShipmentExpense(BaseModel):
    """
    Shipment Expense Model
    Represents expenses related to a shipment (fuel, tolls, etc.)
    """

    __tablename__ = "shipment_expenses"

    # Multi-tenancy
    tenant_id = Column(UUID(as_uuid=True), ForeignKey("tenants.id"), nullable=False)

    # Relationships
    shipment_id = Column(
        UUID(as_uuid=True),
        ForeignKey("shipments.id"),
        nullable=False,
        index=True,
        comment="Related shipment",
    )

    # Expense Details
    expense_type = Column(
        SQLEnum(ExpenseType),
        nullable=False,
        comment="Type of expense",
    )
    amount = Column(Numeric(10, 2), nullable=False, comment="Expense amount")
    currency = Column(String(3), nullable=False, default="USD", comment="Currency")
    expense_date = Column(Date, nullable=False, comment="Date of expense")

    # Details
    description = Column(Text, nullable=True, comment="Expense description")
    location = Column(String(255), nullable=True, comment="Location where expense occurred")
    receipt_url = Column(String(500), nullable=True, comment="URL to receipt/invoice")

    # Relationships
    shipment = relationship("Shipment", back_populates="expenses")

    def __repr__(self):
        return f"<ShipmentExpense {self.expense_type.value} - {self.amount}>"

File: frontend/types/transport.ts
/**
 * Transport Types
 * TypeScript types synchronized with backend Pydantic schemas
 * Source: backend/modules/transport/schemas.py
 */

/**
 * Vehicle type enum
 * Synced with: backend/models/transport.py - VehicleType enum
 */
export enum VehicleType {
  CAR = 'car',
  VAN = 'van',
  TRUCK = 'truck',
  MOTORCYCLE = 'motorcycle',
  OTHER = 'other',
}

/**
 * Vehicle status enum
 * Synced with: backend/models/transport.py - VehicleStatus enum
 */
export enum VehicleStatus {
  ACTIVE = 'active',
  MAINTENANCE = 'maintenance',
  INACTIVE = 'inactive',
}

/**
 * Shipment status enum
 * Synced with: backend/models/transport.py - ShipmentStatus enum
 */
export enum ShipmentStatus {
  PENDING = 'pending',
  IN_TRANSIT = 'in_transit',
  DELIVERED = 'delivered',
  CANCELLED = 'cancelled',
}

/**
 * Shipment expense type enum
 * Synced with: backend/models/transport.py - ExpenseType enum
 */
export enum ShipmentExpenseType {
  FUEL = 'fuel',
  TOLL = 'toll',
  PARKING = 'parking',
  MAINTENANCE = 'maintenance',
  OTHER = 'other',
}

// ============================================================================
// VEHICLE TYPES
// ============================================================================

/**
 * Vehicle response
 * Synced with: VehicleResponse schema
 */
export interface Vehicle {
  id: string
  tenant_id: string
  plate_number: string
  brand: string
  model: string
  year: string | null
  vehicle_type: VehicleType
  status: VehicleStatus
  assigned_driver_id: string | null
  capacity_kg: number | null
  fuel_type: string | null
  fuel_efficiency_km_l: number | null
  last_maintenance_date: string | null
  next_maintenance_date: string | null
  mileage_km: number | null
  notes: string | null
  driver_name: string | null
  shipment_count: number
  created_at: string
  updated_at: string
}

/**
 * Vehicle create request
 * Synced with: VehicleCreate schema
 */
export interface VehicleCreate {
  plate_number: string
  brand: string
  model: string
  year?: string | null
  vehicle_type: VehicleType
  status?: VehicleStatus
  assigned_driver_id?: string | null
  capacity_kg?: number | null
  fuel_type?: string | null
  fuel_efficiency_km_l?: number | null
  last_maintenance_date?: string | null
  next_maintenance_date?: string | null
  mileage_km?: number | null
  notes?: string | null
}

/**
 * Vehicle update request
 * Synced with: VehicleUpdate schema
 */
export interface VehicleUpdate {
  plate_number?: string
  brand?: string
  model?: string
  year?: string | null
  vehicle_type?: VehicleType
  status?: VehicleStatus
  assigned_driver_id?: string | null
  capacity_kg?: number | null
  fuel_type?: string | null
  fuel_efficiency_km_l?: number | null
  last_maintenance_date?: string | null
  next_maintenance_date?: string | null
  mileage_km?: number | null
  notes?: string | null
}

/**
 * Paginated vehicle list response
 * Synced with: VehicleListResponse schema
 */
export interface VehicleListResponse {
  vehicles: Vehicle[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

/**
 * Vehicle summary statistics
 * Synced with: VehicleSummary schema
 */
export interface VehicleSummary {
  total_vehicles: number
  active_vehicles: number
  in_maintenance: number
  inactive_vehicles: number
  total_capacity_kg: number
  avg_fuel_efficiency: number
  vehicles_needing_maintenance: number
}

/**
 * Vehicle filters for list queries
 * Synced with: VehicleFilters schema
 */
export interface VehicleFilters {
  status?: VehicleStatus
  vehicle_type?: VehicleType
  assigned_driver_id?: string
  search?: string
  page?: number
  page_size?: number
}

// ============================================================================
// SHIPMENT EXPENSE TYPES
// ============================================================================

/**
 * Shipment expense response
 * Synced with: ShipmentExpenseResponse schema
 */
export interface ShipmentExpense {
  id: string
  tenant_id: string
  shipment_id: string
  expense_type: ShipmentExpenseType
  amount: number
  currency: string
  expense_date: string
  description: string | null
  location: string | null
  receipt_url: string | null
  created_at: string
  updated_at: string
}

/**
 * Shipment expense create request
 * Synced with: ShipmentExpenseCreate schema
 */
export interface ShipmentExpenseCreate {
  expense_type: ShipmentExpenseType
  amount: number
  currency?: string
  expense_date: string
  description?: string | null
  location?: string | null
  receipt_url?: string | null
}

/**
 * Shipment expense update request
 * Synced with: ShipmentExpenseUpdate schema
 */
export interface ShipmentExpenseUpdate {
  expense_type?: ShipmentExpenseType
  amount?: number
  currency?: string
  expense_date?: string
  description?: string | null
  location?: string | null
  receipt_url?: string | null
}

// ============================================================================
// SHIPMENT TYPES
// ============================================================================

/**
 * Shipment response
 * Synced with: ShipmentResponse schema
 */
export interface Shipment {
  id: string
  tenant_id: string
  shipment_number: string
  client_id: string | null
  vehicle_id: string | null
  driver_id: string | null
  origin_address: string
  origin_city: string
  destination_address: string
  destination_city: string
  scheduled_date: string
  pickup_date: string | null
  delivery_date: string | null
  description: string | null
  weight_kg: number | null
  quantity: number | null
  estimated_distance_km: number | null
  actual_distance_km: number | null
  freight_cost: number
  currency: string
  status: ShipmentStatus
  notes: string | null
  client_name: string | null
  vehicle_plate: string | null
  driver_name: string | null
  total_expenses: number
  expense_count: number
  created_at: string
  updated_at: string
}

/**
 * Shipment with full expense details
 * Synced with: ShipmentWithExpenses schema
 */
export interface ShipmentWithExpenses extends Shipment {
  expenses: ShipmentExpense[]
}

/**
 * Shipment create request
 * Synced with: ShipmentCreate schema
 */
export interface ShipmentCreate {
  shipment_number: string
  client_id?: string | null
  vehicle_id?: string | null
  driver_id?: string | null
  origin_address: string
  origin_city: string
  destination_address: string
  destination_city: string
  scheduled_date: string
  pickup_date?: string | null
  delivery_date?: string | null
  description?: string | null
  weight_kg?: number | null
  quantity?: number | null
  estimated_distance_km?: number | null
  actual_distance_km?: number | null
  freight_cost?: number
  currency?: string
  status?: ShipmentStatus
  notes?: string | null
}

/**
 * Shipment update request
 * Synced with: ShipmentUpdate schema
 */
export interface ShipmentUpdate {
  shipment_number?: string
  client_id?: string | null
  vehicle_id?: string | null
  driver_id?: string | null
  origin_address?: string
  origin_city?: string
  destination_address?: string
  destination_city?: string
  scheduled_date?: string
  pickup_date?: string | null
  delivery_date?: string | null
  description?: string | null
  weight_kg?: number | null
  quantity?: number | null
  estimated_distance_km?: number | null
  actual_distance_km?: number | null
  freight_cost?: number
  currency?: string
  status?: ShipmentStatus
  notes?: string | null
}

/**
 * Paginated shipment list response
 * Synced with: ShipmentListResponse schema
 */
export interface ShipmentListResponse {
  shipments: Shipment[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

/**
 * Shipment summary statistics
 * Synced with: ShipmentSummary schema
 */
export interface ShipmentSummary {
  total_shipments: number
  pending_shipments: number
  in_transit_shipments: number
  delivered_shipments: number
  cancelled_shipments: number
  total_revenue: number
  total_expenses: number
  net_profit: number
  avg_distance_km: number
  total_distance_km: number
}

/**
 * Shipment filters for list queries
 * Synced with: ShipmentFilters schema
 */
export interface ShipmentFilters {
  status?: ShipmentStatus
  client_id?: string
  vehicle_id?: string
  driver_id?: string
  origin_city?: string
  destination_city?: string
  scheduled_date_from?: string
  scheduled_date_to?: string
  delivery_date_from?: string
  delivery_date_to?: string
  search?: string
  page?: number
  page_size?: number
}

// ============================================================================
// HELPER TYPES
// ============================================================================

/**
 * Helper type for status badge colors
 */
export const VEHICLE_STATUS_COLORS = {
  [VehicleStatus.ACTIVE]: 'green',
  [VehicleStatus.MAINTENANCE]: 'yellow',
  [VehicleStatus.INACTIVE]: 'gray',
} as const

/**
 * Helper type for status badge colors
 */
export const SHIPMENT_STATUS_COLORS = {
  [ShipmentStatus.PENDING]: 'gray',
  [ShipmentStatus.IN_TRANSIT]: 'blue',
  [ShipmentStatus.DELIVERED]: 'green',
  [ShipmentStatus.CANCELLED]: 'red',
} as const

/**
 * Helper type for status badge labels
 */
export const VEHICLE_STATUS_LABELS = {
  [VehicleStatus.ACTIVE]: 'Activo',
  [VehicleStatus.MAINTENANCE]: 'Mantenimiento',
  [VehicleStatus.INACTIVE]: 'Inactivo',
} as const

/**
 * Helper type for status badge labels
 */
export const SHIPMENT_STATUS_LABELS = {
  [ShipmentStatus.PENDING]: 'Pendiente',
  [ShipmentStatus.IN_TRANSIT]: 'En Tránsito',
  [ShipmentStatus.DELIVERED]: 'Entregado',
  [ShipmentStatus.CANCELLED]: 'Cancelado',
} as const

/**
 * Helper type for expense type labels
 */
export const EXPENSE_TYPE_LABELS = {
  [ShipmentExpenseType.FUEL]: 'Combustible',
  [ShipmentExpenseType.TOLL]: 'Peaje',
  [ShipmentExpenseType.PARKING]: 'Estacionamiento',
  [ShipmentExpenseType.MAINTENANCE]: 'Mantenimiento',
  [ShipmentExpenseType.OTHER]: 'Otro',
} as const
